File: api/src/grpc/discussion.rs
use core_services::services::base::Service;
use schema::devlog::devblog::rpc::devblog_discussion_service_server::DevblogDiscussionService;
use schema::devlog::devblog::rpc::{
    GetDiscussionsRequest,
    GetDiscussionsResponse,
    NewDiscussionRequest,
    NewDiscussionResponse,
};
use schema::devlog::entities::User;
use tonic::{Request, Response, Result, Status};

use crate::di::ApiDependenciesInjection;
use crate::services::discussion::{GetListDiscussionsParam, NewDiscussionParams};

#[derive(Clone,)]
pub struct DiscussionGrpcService {
    pub(crate) di: &'static ApiDependenciesInjection,
}

#[async_trait::async_trait]
impl DevblogDiscussionService for DiscussionGrpcService {
    async fn new_discussion(
        &self,
        request: Request<NewDiscussionRequest,>,
    ) -> Result<Response<NewDiscussionResponse,>, Status,> {
        let user: &User = request.extensions().get::<User>().ok_or(Status::unauthenticated("You're not authorize",),)?;
        let request = request.get_ref();
        let service = self.di.new_discussion_service();

        let params = NewDiscussionParams {
            user,
            discussion: request.new_discussion.as_ref().expect("new_discussion must be defined",),
            post_id: request.post_id.as_ref().expect("Post_id must be defined",),
        };

        service.execute(params,).await?;
        let response = NewDiscussionResponse {};
        Ok(Response::new(response,),)
    }

    async fn get_discussions(
        &self,
        request: Request<GetDiscussionsRequest,>,
    ) -> Result<Response<GetDiscussionsResponse,>, Status,> {
        let request = request.get_ref();
        let service = self.di.list_discussion_service();
        let param = GetListDiscussionsParam {
            paging: request.paging.as_ref().unwrap().clone(),
            post_id: request.post_id.clone().unwrap(),
        };

        let result = service.execute(param,).await?;
        Ok(Response::new(GetDiscussionsResponse {
            discussions: result.discussions,
            paging: Some(result.paging,),
        },),)
    }
}

File: web/src/app/services/discussion.ts
import {
  PostId,
  NewDiscussionRequest,
  Discussion,
  GetDiscussionsRequest,
  Paging, GetDiscussionsResponse, DevblogDiscussionService,
} from "@devlog/schema-ts"

import gRPCClientBase from "./base"

export default class DiscussionService extends gRPCClientBase<typeof DevblogDiscussionService> {
  constructor() {
    super(DevblogDiscussionService)
  }

  async newDiscussion(content: string, postTitle: string): Promise<boolean> {
    if (!content) {
      throw "Empty message !!"
    }

    const request = new NewDiscussionRequest()
    const postId = new PostId()
    postId.title = postTitle

    const newDiscussion = new Discussion()
    newDiscussion.content = content

    request.newDiscussion = newDiscussion

    try {
      request.postId = postId
    } catch (ignore) {
      console.log(ignore)
    }

    await this.client.new_discussion(request, { headers: this.getHeader() })
    return true
  }

  async getDiscussions(page: Paging, title: string): Promise<{ discussions: Discussion[], paging: Paging }> {
    if (page.page < 1) throw 'The page must be a positive number'
    if (page.rowsPerPage < 1) throw 'The rowsPerPage must be a positive number'

    const request = new GetDiscussionsRequest()
    request.paging = page
    request.postId = new PostId({
      title
    })

    const response = await this.client.get_discussions(request, { headers: this.getHeader() }) as GetDiscussionsResponse

    const discussionsList = response.discussions.map(discussion => discussion)

    return {
      discussions: discussionsList,
      paging: response.paging!,
    }
  }
}
